src/ORM/Association/HasMany.php

```diff
@@ -406,25 +410,31 @@ function ($v) {
      * @param array $foreignKey array of foreign key properties
      * @param Table $target The associated table
      * @param array $conditions The conditions that specifies what are the objects to be unlinked
-     * @return void
+     * @return bool success
      */
     protected function _unlink(array $foreignKey, Table $target, array $conditions = [])
     {
+        $ok = true;
         $mustBeDependent = (!$this->_foreignKeyAcceptsNull($target, $foreignKey) || $this->dependent());
-        if ($mustBeDependent) {
-            if ($this->_cascadeCallbacks) {
-                $query = $this->find('all')->where($conditions);
-                foreach ($query as $assoc) {
-                    $target->delete($assoc);
+        $persistedEntitiesExist = $this->exists($conditions);
+
+        if ($persistedEntitiesExist) {
+            if ($mustBeDependent) {
+                if ($this->_cascadeCallbacks) {
+                    $query = $this->find('all')->where($conditions);
+                    foreach ($query as $assoc) {
+                        $ok = $ok && $target->delete($assoc);
+                    }
+                } else {
+                    $ok = ($target->deleteAll($conditions) > 0);
                 }
             } else {
-                $target->deleteAll($conditions);
-            }
-        } else {
-            $updateFields = array_fill_keys($foreignKey, null);
-            $target->updateAll($updateFields, $conditions);
+                $updateFields = array_fill_keys($foreignKey, null);
+                $ok = $target->updateAll($updateFields, $conditions);
```

Comment:
When I write methods I use to write only one return at the end and during the method all that changes is the value of the variable to be returned. Do you think it is better to have multiple returns instead or it doesn't really matter?